support/mesh/mesh_channel_core/src/oneshot.rs

```diff
@@ -0,0 +1,689 @@
+                }
+                SlotState::ReceiverRemote(port, _) => port,
+                SlotState::Done => Port::new_pair().0,
+                SlotState::Sent(_) | SlotState::SenderRemote { .. } => unreachable!(),
+            }
+        }
+        into_port(self, decode_message::<T>)
+    }
+
+    fn from_port<T: 'static + MeshField + Send>(port: Port) -> Self {
+        fn from_port(port: Port, encode: EncodeFn) -> OneshotSenderCore {
+            let slot = Arc::new(Slot(Mutex::new(SlotState::ReceiverRemote(port, encode))));
+            OneshotSenderCore(slot)
+        }
+        from_port(port, encode_message::<T>)
+    }
+}
+
+/// The receiving half of a channel returned by [`oneshot`].
+///
+/// A value is received by `poll`ing or `await`ing the channel.
+//
+// Note that the `PhantomData` here is necessary to ensure `Send/Sync` traits
+// are only implemented when `T` is `Send`, since the `OneshotReceiverCore` is
+// always `Send+Sync`. This behavior is verified in the unit tests.
+pub struct OneshotReceiver<T>(
+    ManuallyDrop<OneshotReceiverCore>,
+    PhantomData<Arc<Mutex<T>>>,
+);
+
+impl<T> Debug for OneshotReceiver<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Debug::fmt(&self.0, f)
+    }
+}
+
+impl<T> OneshotReceiver<T> {
+    fn poll_recv(&mut self, cx: &mut Context<'_>) -> Poll<Result<T, RecvError>> {
+        // SAFETY: the slot is of type `T`.
+        let v = unsafe { ready!(self.0.poll_recv(cx))? };
+        Ok(*v).into()
+    }
+
+    fn into_core(self) -> OneshotReceiverCore {
+        let Self(ref core, _) = *ManuallyDrop::new(self);
+        // SAFETY: `core` is not dropped.
+        unsafe { <*const _>::read(&**core) }
+    }
+}
+
+impl<T> Drop for OneshotReceiver<T> {
+    fn drop(&mut self) {
+        // SAFETY: the core is not dropped and will never be used again.
+        let core = unsafe { ManuallyDrop::take(&mut self.0) };
+        // SAFETY: the slot is of type `T`.
+        unsafe { core.drop::<T>() };
+    }
+}
+
+impl<T> Future for OneshotReceiver<T> {
+    type Output = Result<T, RecvError>;
+
+    fn poll(self: std::pin::Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.get_mut().poll_recv(cx)
+    }
+}
+
+impl<T: 'static + MeshField + Send> DefaultEncoding for OneshotReceiver<T> {
+    type Encoding = PortField;
+}
+
+impl<T: 'static + MeshField + Send> From<OneshotReceiver<T>> for Port {
+    fn from(receiver: OneshotReceiver<T>) -> Self {
+        // SAFETY: the slot is of type `T`.
+        unsafe { receiver.into_core().into_port::<T>() }
+    }
+}
+
+impl<T: 'static + MeshField + Send> From<Port> for OneshotReceiver<T> {
+    fn from(port: Port) -> Self {
+        Self(
+            ManuallyDrop::new(OneshotReceiverCore::from_port::<T>(port)),
+            PhantomData,
+        )
+    }
+}
+
+#[derive(Debug)]
+struct OneshotReceiverCore {
+    slot: Arc<Slot>,
+    port: Option<PortWithHandler<SlotHandler>>,
```

Comment:
is there no way to squeeze the port into the `slot` itself, so local oneshot receivers don't have to pay the cost of this always-None field?

Reply:
A little too tricky for v1 due to our use of `Arc`. I'll add a comment to improve this later.